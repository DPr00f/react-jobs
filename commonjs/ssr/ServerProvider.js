'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ClientProvider = require('./ClientProvider');

var _ClientProvider2 = _interopRequireDefault(_ClientProvider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } // eslint-disable-next-line import/no-extraneous-dependencies


var ServerProvider = function (_Component) {
  _inherits(ServerProvider, _Component);

  function ServerProvider(props) {
    _classCallCheck(this, ServerProvider);

    var _this = _possibleConstructorReturn(this, (ServerProvider.__proto__ || Object.getPrototypeOf(ServerProvider)).call(this, props));

    if (props.runJobsExecContext) {
      _this.jobStates = props.runJobsExecContext.getState().jobsState;
    } else {
      _this.jobStates = {};
    }
    return _this;
  }

  _createClass(ServerProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      var _this2 = this;

      var context = {
        reactJobsServer: {
          registerJobState: function registerJobState(jobID, jobState) {
            _this2.jobStates[jobID] = jobState;
            if (_this2.context.runJobsExecContext) {
              _this2.context.runJobsExecContext.registerJobState(jobID, jobState);
            }
          },
          getJobState: function getJobState(jobID) {
            return _this2.jobStates[jobID];
          }
        }
      };

      return context;
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _ClientProvider2.default,
        null,
        this.props.children
      );
    }
  }]);

  return ServerProvider;
}(_react.Component);

ServerProvider.contextTypes = {
  reactJobsRunner: _react.PropTypes.object
};
ServerProvider.childContextTypes = {
  reactJobsServer: _react.PropTypes.object.isRequired
};

exports.default = ServerProvider;