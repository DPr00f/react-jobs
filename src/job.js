/* @flow */

import React, { Component, PropTypes } from 'react';
import { getDisplayName, isPromise } from './utils';
import type {
  ProviderContext,
  JobState,
} from './server/types';

type Work = (props : Object) => any;

type State = JobState & { executingJob?: Promise<any> };

type Props = {
  jobID?: number,
  [key: string]: any,
};

export default function job(work : Work) {
  if (typeof work !== 'function') {
    throw new Error('You must provide a function to a react-jobs job declaration.');
  }

  return function WrapComponentWithJob(WrappedComponent : Function) {
    class ComponentWithJob extends Component {
      props: Props;
      state: State;

      constructor(props : Props, context : ProviderContext) {
        super(props, context);
        this.state = { inProgress: false };
      }

      componentWillMount() {
        if (this.context.reactJobsClient) {
          const ssrRehydrateState = this.context
            .reactJobsClient
            .popJobRehydrationForSRR(this.props.jobID);
          if (ssrRehydrateState) {
            this.setState(ssrRehydrateState);
            return;
          }
        }

        if (this.context.reactJobsServer) {
          const jobState = this.context.reactJobsServer.getJobState(this.props.jobID);
          if (jobState) {
            this.setState(jobState);
            return;
          }
        }

        const context : ProviderContext = this.context;
        const workResult = work(this.props);

        if (isPromise(workResult)) {
          workResult
            .then((result) => {
              this.setState({ inProgress: false, result });
              return result;
            })
            .catch((error) => {
              console.warn('Job failed:\n', error); // eslint-disable-line no-console
              this.setState({ inProgress: false, error });
              return 'ouchy';
            })
            .then(() => {
              if (this.props.jobID && context.reactJobsServer) {
                context.reactJobsServer.registerJobState(
                  this.props.jobID,
                  this.getJobState(),
                );
              }
            });

          // Asynchronous result.
          this.setState({ inProgress: true, executingJob: workResult });
        } else {
          // Synchronous result.
          this.setState({ result: workResult });
        }
      }

      getExecutingJob() {
        if (!this.state.executingJob) {
          return undefined;
        }
        return this.state.executingJob;
      }

      getJobState() : JobState {
        const { inProgress, result, error } = this.state;
        return { inProgress, result, error };
      }

      render() {
        const jobState = this.getJobState();
        return <WrappedComponent {...this.props} job={jobState} />;
      }
    }
    ComponentWithJob.displayName = `${getDisplayName(WrappedComponent)}WithJob`;
    ComponentWithJob.contextTypes = {
      reactJobsClient: PropTypes.object,
      reactJobsServer: PropTypes.object,
    };
    return ComponentWithJob;
  };
}
