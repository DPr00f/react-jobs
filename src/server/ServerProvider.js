/* @flow */

import { Children, Component, Element, PropTypes } from 'react';
import type { ServerProviderContext, RenderContext } from './types';

type React$Element = Element<*>;

type Props = {
  children?: React$Element,
  renderContext: RenderContext,
};

let currentjobID = 0;
const jobResults = {};

class ServerProvider extends Component {
  props: Props;

  getChildContext() {
    const context : ServerProviderContext = {
      reactJobsServer: {
        nextJobID: () => {
          currentjobID += 1;

          if (this.props.renderContext) {
            this.props.renderContext.setCurrentJobID(currentjobID);
          }

          return currentjobID;
        },
        registerJobResults: (jobID: number, results: any) => {
          jobResults[jobID] = results;
          if (this.props.renderContext) {
            this.props.renderContext.registerResolvedJob(jobID);
          }
        },
        getJobResults: (jobID: number) => jobResults[jobID],
      },
    };

    return context;
  }

  render() {
    return Children.only(this.props.children);
  }
}

ServerProvider.childContextTypes = {
  reactJobs: PropTypes.object.isRequired,
};

export default ServerProvider;
