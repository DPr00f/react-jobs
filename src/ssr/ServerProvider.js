/*  */

// eslint-disable-next-line import/no-extraneous-dependencies
import React, { Component, PropTypes } from 'react'
import ClientProvider from './ClientProvider'

class ServerProvider extends Component {
  constructor(props) {
    super(props)
    if (props.runJobsExecContext) {
      this.jobStates = props.runJobsExecContext.getState().jobsState
    } else {
      this.jobStates = {}
    }
  }

  getChildContext() {
    const context = {
      reactJobsServer: {
        registerJobState: (jobID, jobState) => {
          this.jobStates[jobID] = jobState
          if (this.context.runJobsExecContext) {
            this.context.runJobsExecContext.registerJobState(jobID, jobState)
          }
        },
        getJobState: jobID => this.jobStates[jobID],
      },
    }

    return context
  }

  render() {
    return (
      <ClientProvider>
        {this.props.children}
      </ClientProvider>
    )
  }
}

ServerProvider.contextTypes = {
  reactJobsRunner: PropTypes.object,
}
ServerProvider.childContextTypes = {
  reactJobsServer: PropTypes.object.isRequired,
}

export default ServerProvider
