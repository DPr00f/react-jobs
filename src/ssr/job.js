/* @flow */
// @see https://github.com/facebook/react/issues/1137

import React, { PropTypes } from 'react';
import { getDisplayName } from '../utils';
import clientOnlyJob from '../job';
import type { ReactJobsClientContext, ReactJobsServerContext } from './types';

type Context = {
  reactJobsClient: ReactJobsClientContext,
  reactJobsServer?: ReactJobsServerContext,
};

const getInitialState = (context, jobID) => {
  const { reactJobsClient, reactJobsServer } = context;

  if (reactJobsServer) {
    // Running on the server. If the jobs have been executed by the runJobs
    // helper then we will get back a result here.
    const jobState = reactJobsServer.getJobState(jobID);
    if (jobState) {
      return jobState;
    }
  } else {
    // Otherwise we have to be on the browser. Check if there is any
    // rehydration state available for this job.
    const ssrRehydrateState = reactJobsClient.popJobRehydrationForSRR(jobID);
    if (ssrRehydrateState) {
      return ssrRehydrateState;
    }
  }

  return undefined;
};

const job = (work : any) => (Component : Function) => {
  let jobIDHandle = null;

  function withSSRBehaviour(WrappedComponent) {
    const ComponentWithJobID = (props : Object, context : Context) => {
      const { reactJobsClient, reactJobsServer } = context;

      // Every SSR job should have the client attached in order to provide
      // the identifiers for the components.
      if (!reactJobsClient) {
        throw new Error('You are using an "ssr/job" without having wrapped your component render with the "runJobs" or "rehydrateJobs" helpers.  Please refer to the docs for more details.');
      }

      // Establish a unique identifier for this job instance.
      const getJobID = () => {
        if (!jobIDHandle) {
          jobIDHandle = reactJobsClient.nextJobID();
        }
        return jobIDHandle;
      };
      const jobID = getJobID();

      // Determine if we have an state from server side processing or
      // client side rehdration.
      const initialState = getInitialState(context, jobID);

      // When on a server we want to call back and register the result of the
      // work so that the server can hydrate the app appropriately as well
      // as provide the initial state to be sent to the browser.
      const onJobProcessed = reactJobsServer
        ? jobsState => reactJobsServer.registerJobState(jobID, jobsState)
        : null;

      return (
        <WrappedComponent
          {...props}
          jobInitState={initialState}
          onJobProcessed={onJobProcessed}
        />
      );
    };
    ComponentWithJobID.displayName = `${getDisplayName(WrappedComponent)}WithJobID`;
    ComponentWithJobID.contextTypes = {
      reactJobsClient: PropTypes.object.isRequired,
      reactJobsServer: PropTypes.object,
    };
    return ComponentWithJobID;
  }

  // We wrap the standard job implementation with our SSR behaviour.
  return withSSRBehaviour(clientOnlyJob(work)(Component));
};

export default job;
