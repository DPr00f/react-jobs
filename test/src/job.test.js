/* @flow */

import React from 'react';
import { mount } from 'enzyme';
import { Foo, resolveAfter, rejectAfter, warningsAsErrors } from '../__helpers__';
import ServerProvider from '../../src/server/ServerProvider';

// Under Test.
import job from '../../src/job';

const workTime = 10; // ms

describe('job()', () => {
  warningsAsErrors();

  describe('options', () => {
    it('returns a function', () => {
      const actual = typeof job(() => undefined);
      const expected = 'function';
      expect(actual).toEqual(expected);
    });

    it('should throws if no work is provided', () => {
      // $FlowIgnore: we expect this to flow error
      expect(() => job()).toThrowError(/provide a function to a react-jobs/);
    });

    it('should throws if the work is invalid', () => {
      // $FlowIgnore: we expect this to flow error
      expect(() => job(1)).toThrowError(/provide a function to a react-jobs/);
    });
  });

  describe('higher order component', () => {
    const hoc = job(() => resolveAfter(1));
    const Actual = hoc(Foo);

    it('should return a renderable component', () => {
      expect(() => mount(<Actual />)).not.toThrowError();
    });
  });

  describe('rendering', () => {
    it('should set the "result" immediately if the work does not return a promise', () => {
      const FooWithJob = job(() => 'bob')(Foo);
      expect(mount(<ServerProvider><FooWithJob /></ServerProvider>)).toMatchSnapshot();
    });

    it('should provide the props to the work function', () => {
      const expected = { foo: 'bar', baz: 'qux' };
      let actual;
      const FooWithJob = job((props) => { actual = props; })(Foo);
      mount(<FooWithJob {...expected} />);
      expect(actual).toMatchObject(expected);
    });

    it('should set "inProgress" when processing work', () => {
      const FooWithJob = job(() => resolveAfter(workTime))(Foo);
      const actual = mount(<FooWithJob />).find(Foo).props();
      const expected = { job: { inProgress: true } };
      expect(actual).toMatchObject(expected);
    });

    it('should set "result" when work completes successfully', () => {
      const FooWithJob = job(() => resolveAfter(workTime, 'result'))(Foo);
      const renderWrapper = mount(<FooWithJob />);
      // Allow enough time for work to complete
      return resolveAfter(workTime + 5).then(() => {
        const actual = renderWrapper.find(Foo).props();
        const expected = { job: { inProgress: false, result: 'result' } };
        expect(actual).toMatchObject(expected);
      });
    });

    it('should set "error" when work fails', () => {
      const FooWithJob = job(() => rejectAfter(workTime, 'error'))(Foo);
      const renderWrapper = mount(<FooWithJob />);
      // Allow enough time for work to complete
      return resolveAfter(workTime + 5).then(() => {
        const actual = renderWrapper.find(Foo).props();
        const expected = { job: { inProgress: false, error: 'error' } };
        expect(actual).toMatchObject(expected);
      });
    });
  });
});
