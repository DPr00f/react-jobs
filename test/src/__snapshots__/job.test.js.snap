exports[`job() rendering should set the "result" immediately if the work does not return a promise 1`] = `
<ServerProvider>
  <FooWithJobWithJobID>
    <FooWithJob
      jobID={1}>
      <Foo
        job={
          Object {
            "error": undefined,
            "inProgress": false,
            "result": "bob",
          }
        }
        jobID={1}>
        <div>
          bob
        </div>
      </Foo>
    </FooWithJob>
  </FooWithJobWithJobID>
</ServerProvider>
`;
